dashboard/src/components/auth/LoginForm.tsx

```diff
@@ -70,20 +111,48 @@ export default function LoginForm() {
             onChange={(e) => setPassword(e.target.value)}
             className='border-input bg-background text-foreground focus:ring-ring placeholder:text-muted-foreground w-full rounded-md border px-3 py-2 focus:border-transparent focus:ring-2 focus:outline-none'
             placeholder='Enter your password'
+            disabled={isDialogOpen}
           />
         </div>
       </div>
 
       <div>
         <button
           type='submit'
-          disabled={isPending}
+          disabled={isPending || isDialogOpen}
           className='text-primary-foreground bg-primary hover:bg-primary/90 focus:ring-ring flex w-full justify-center rounded-md border border-transparent px-4 py-2 text-sm font-medium transition-colors focus:ring-2 focus:ring-offset-2 focus:outline-none disabled:cursor-not-allowed disabled:opacity-50'
         >
-          {isPending ? 'Signing in...' : 'Sign in'}
+          {isPending || isDialogOpen ? 'Signing in...' : 'Sign in'}
         </button>
       </div>
 
+      <AlertDialog open={isDialogOpen} onOpenChange={setIsDialogOpen}>
+        <AlertDialogContent className='max-h-[90vh] w-80 overflow-y-auto'>
+          <AlertDialogHeader>
+            <AlertDialogTitle>Two-factor authentication</AlertDialogTitle>
+            <AlertDialogDescription>
+              Enter the verification code from your authenticator app.
+            </AlertDialogDescription>
+          </AlertDialogHeader>
+          <Error />
+          {isPending ? (
+            <Spinner className='m-auto' />
+          ) : (
+            <OtpInput
+              value={totp}
+              onValueChange={setTotp}
+              disabled={isPending}
+              ref={totpInputRef}
+              autoSubmit
+              form='login'
+            />
+          )}
+          <AlertDialogFooter>
+            <AlertDialogCancel disabled={isPending}>Cancel</AlertDialogCancel>
+          </AlertDialogFooter>
+        </AlertDialogContent>
+      </AlertDialog>
```

Comment:
Suggestion: It works super nice! However, I would move this into a separate page, such that you get redirected from the `/signin` page to `/mfa` for instance. That would both help simplifying this page and flow, optimize for accessibility, and show the correct "Save password" at the correct timing. Currently, my password manager only asks to save the password after the TOTP has been filled in😄

Reply:
Given how this is handled via `next-auth`, the `email`, `password` and `totp` have to be passed together to `verifyCredentials`, and it seemed challenging to achieve this when redirecting the user to another page. I'd have to save the password somewhere as plain text to resubmit it, or add some logic to the session to be in a state mid-way between not being signed in and being fully signed in

Imho, the password manager's current behavior makes sense, because it should prompt to save the credentials after a successful login attempt, i.e. after entering a valid TOTP. Imagine a scenario where a hacker manages to figure out a user's credentials, but fails to provide a valid TOTP - why would they save the credentials in this case? 🤔